h2o-core/src/main/java/hex/grid/GridSearch.java

```diff
@@ -118,8 +117,7 @@ private GridSearch(Key<Grid> gkey, HyperSpaceWalker<MP, ?> hyperSpaceWalker, fin
     HyperSpaceWalker.HyperSpaceIterator<MP> it = _hyperSpaceWalker.iterator();
     long gridWork=0;
     if (gridSize > 0) {//if total grid space is known, walk it all and count up models to be built (not subject to time-based or converge-based early stopping)
-      int count=0;
-      while (it.hasNext(model) && (it.max_models() > 0 && count++ < it.max_models())) { //only walk the first max_models models, if specified
```

Comment:
Please keep in mind grid search restarts, a similar thing was already discussed in https://github.com/h2oai/h2o-3/pull/4191#discussion_r370252892

Is the behavior the same as before?

Reply:
It is just a refactoring, I rely on existing tests to check that the legacy behaviour is not broken.
If we have expectations that are nowhere in tests, then they are barely features.

Will check though, but really, if we don't trust our tests, then we can't touch anything.
Even the expectations in the discussion linked above don't seem always clear to everyone.

Reply:
@michalkurka FYI the code example you linked above still produces 2 models for each grid, as expected.
I got confused by you naming this "grid restart" as I though you wanted some kind of `pause/continue` behaviour that we're not currently supporting afaik.

Reply:
we are supporting grid pause & restart - see the linked discussion

Reply:
@michalkurka  I was really surprised by this, so tried and confirm that it resumes grids.
First thought that it was because the `walker` was keeping the state of the iterator and then reused in the `resume` run, but it doesn't seem to happen this way: on reruns, there's actually a new walker created each time, but when building model, there is an "optimization": grid search looks up in the DKV for a model with the same params, and if found, it adds it to the grid.
So, what I guess is happening (I tried with a seed to be sure it doesn't rebuild the same models multiple times), is that it is still iterating from the start, but when finds the previous model, it re-adds it to the grid with its old key.
See `GridSearch#buildModel` for details.

I'll try it with this branch to confirm, but if my hypothesis is correct, it should still work.

Reply:
Will also try to find an existing test for this and create one if needed.

Reply:
Yes, it is designed to be restartable. It works by looking up the models and checking checksums on parameters and frames. This way it can identify they belong to the grid and avoid rebuilding them again.

Reply:
(the issue we discussed with Andrey in the other PR is that max_models should be respected - it shouldn't exceed the original max_models specification)

Reply:
to test this, I created a method on Python to cancel jobs (needed this for an AutoML test): you just need to call it in a subprocess after x seconds, restart the grid, recancel, compare the grid state.